spec/unit/pops/types/p_sensitive_type_spec.rb

```diff
@@ -0,0 +1,127 @@
+          notice(Sensitive < Sensitive)
+          notice(Sensitive > Sensitive)
+        CODE
+        expect(eval_and_collect_notices(code)).to eq(['true', 'true', 'false', 'false'])
+      end
+
+      context "when parameterized" do
+        it 'is equal other types with the same parameterization' do
+          code = <<-CODE
+            notice(Sensitive[String] == Sensitive[String])
+            notice(Sensitive[Numeric] != Sensitive[Integer])
+          CODE
+          expect(eval_and_collect_notices(code)).to eq(['true', 'true'])
+        end
+
+        it 'orders parameterized types based on the type system hierarchy' do
+          code = <<-CODE
+            notice(Sensitive[Numeric] > Sensitive[Integer])
+            notice(Sensitive[Numeric] < Sensitive[Integer])
+          CODE
+          expect(eval_and_collect_notices(code)).to eq(['true', 'false'])
+        end
+
+        it 'does not order incomparable parameterized types' do
+          code = <<-CODE
+            notice(Sensitive[String] < Sensitive[Integer])
+            notice(Sensitive[String] > Sensitive[Integer])
+          CODE
+          expect(eval_and_collect_notices(code)).to eq(['false', 'false'])
+        end
+
+        it 'generalizes passed types to prevent information leakage' do
+          code =<<-CODE
+            $it = String[7, 7]
+            $st = Sensitive[$it]
+            notice(type($st))
+          CODE
+          expect(eval_and_collect_notices(code)).to eq(['Type[Sensitive[String]]'])
+        end
+      end
+    end
+  end
+
+  context 'a Sensitive instance' do
+    it 'can be created from a string and does not leak its contents' do
+      code =<<-CODE
+        $o = Sensitive("hunter2")
+        notice($o)
+        notice(type($o))
+      CODE
+      expect(eval_and_collect_notices(code)).to eq(['Sensitive [value redacted]', 'Sensitive[String]'])
+    end
+
+    it 'matches the appropriate parameterized type' do
+      code =<<-CODE
+        $o = Sensitive("hunter2")
+        notice(assert_type(Sensitive[String], $o))
+        notice(assert_type(Sensitive[String[7, 7]], $o))
+      CODE
+      expect(eval_and_collect_notices(code)).to eq(['Sensitive [value redacted]', 'Sensitive [value redacted]'])
+    end
+
+    it 'does not match an inappropriate parameterized type' do
+      code =<<-CODE
+        $o = Sensitive("hunter2")
+        notice(assert_type(Sensitive[Integer], $o) |$expected, $actual| {
+          "$expected != $actual"
+        })
+      CODE
+      expect(eval_and_collect_notices(code)).to eq(['Sensitive[Integer] != Sensitive[String]'])
+    end
+
+    it 'can be created from another sensitive instance ' do
+      code =<<-CODE
+        $o = Sensitive("hunter2")
+        $x = Sensitive($o)
+        notice(assert_type(Sensitive[String], $x))
+      CODE
+      expect(eval_and_collect_notices(code)).to eq(['Sensitive [value redacted]'])
+    end
+  end
+
+  it "enforces wrapped type constraints" do
```

Comment:
I don't think that Redacted[T] will play nice here; we're effectively redacting integers and the work that it'll take to allow Redacted[Integer] to be used where Integers once were seems daunting. I had envisioned that Redacted[T] would be used in reports and catalogs and wouldn't be used too much deep within the internals of Puppet. Moreover I meant to use those types to cleanly fail when someone tried to unwrap redacted information; in this case we don't want to blow up if someone tries to read this information - I just want that information elided.